xmodule/docs/decisions/0004-evaluate-wasm-safe-exec.rst

```diff
@@ -0,0 +1,107 @@
+
+Maintenance
+-----------
+
+Content lasts longer than software, and authors don't expect to have to update their scoring code. When Python 2.7 was approaching its end-of-life, it was a large effort to get people to test their content with Python 3. Some of that content is probably forever broken. While this is an extreme example, there will be smaller breakages over time–libraries that get upgraded because the old versions are no longer available, Python versions that get updated because operating systems stop offering end-of-life versions, etc.
+
+The big advantage Wasm provides here is that it can treat these software packages as *data*. If we have a Wasm binary file that holds a compiled version of Python and all relevant packages, then we no longer have to worry about upgrades or whether we can offer exactly those libraries in the future. We just keep the same binary around going forward.
+
+We could still offer upgraded packages for new capabilities, but these would be purely opt-in. You could specify which system you wanted to use in your problem and have confidence that the content would never require updating. So you could specify that this particular problem should use the "python-sci-2024" stack (which maps to a stable Python+libraries release compiled to Wasm), and two years from now, you might make another problem that uses the "python-sci-2026" stack.
+
+Flexibility of Additional Runtimes
+----------------------------------
+
+When the sandboxed runtime environment is just a Wasm file, it makes it much easier for us to support multiple runtimes with the same service and add new ones as necessary. These runtimes can even be in different languages.
+
+Operations
+----------
+
+The Wasmtime library gives a more precise accounting of the resources being used by allotting computation budget "fuel" and measuring how much a given Wasm program has used. This is more precise than the limits that AppArmor uses, which can sometimes result in problems that work fine most of the time but start failing under elevated system load. The fact that we have precise measurements of how much fuel is being used also means that we can surface this information to content authors, so that they can be aware of the problem.
+
+Cross-Platform Support
+----------------------
+
+AppArmor only works in Linux, and cannot run on Linux guest Docker containers on a macOS host machine. This makes it difficult for developers running macOS to develop and test features related to sandboxed code execution. WebAssembly can run on any major operating system.
+
+Decisions
+*********
+
+#. We will create a new library in a new repo (``wasm-safe-exec``) for safely executing instructor-authored code using WebAssembly as the execution mechanism.
+
+   * The library will use ``wasmtime-py`` for Wasm code execution.
+   * The library must be runnable in edx-platform, meaning that it still needs to support Python 3.8 until edx-platform has been upgraded to 3.11.
```

Comment:
[This page](https://component-model.bytecodealliance.org/language-support/python.html#running-components-from-python-applications) seems to imply that wasmtime-py versions that support WebAssembly Component Model require Python 3.11, which may be a problem ?

Reply:
I think what it means is that to create the component requires Python 3.11 (since that's the first version that officially builds to Wasm), but I think the call should work from 3.8 via the `wasmtime-py` lib. I'll try to squeeze some time to set that up and verify that. But if it really is 3.11 only, it shouldn't be too terrible, since we're planning to upgrade edx-platform to that after the Django 4.2 upgrade is complete.

Reply:
... though if you want to try that out and add a note here on your findings, that would be great. 😄